### PR TITLE
Stack watermarking

### PR DESCRIPTION
**Pull request by martinbudden**
@digitalentity , here's the stack watermarking code. I've just tested it on a CJMCU - it's probably best if you independently test it on a separate target before you merge.

Althought it's good to have this, I don't think stack overflow is our problem.

### CODE REVIEW

**Comment by digitalentity** [comment]
I'll test it once I have the time and merge. I don't think attack is our
problem as well. My best guess would be an array index out of bounds,
although I'm not sure where to look.

File: src/main/drivers/stack_check.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * RAM is origin 0x20000000 length 128K that is:
+ * 0x20000000 to 0x20020000
+ *
+ */
+void taskStackCheck(void)
+{
+    char * const stackHighMem = &_estack;
+    const uint32_t stackSize = (uint32_t)&_Min_Stack_Size;
+    char * const stackLowMem = stackHighMem - stackSize;
+    const char * const stackCurrent = (char *)&stackLowMem;
```
Comment (digitalentity) [inline src/main/drivers/stack_check.c:49]: This needs to be the deepest point in stack, currently the check is likely to always fail.

Comment (martinbudden) [reply src/main/drivers/stack_check.c:49]: Not sure what you mean. `p` starts its loop at `stackLowMem`.

Comment (digitalentity) [reply src/main/drivers/stack_check.c:49]: Oh, I see, we are not detecting stack overflow, we are tracking stack usage. Understood. Shouldn't we record stack usage into some system-wide status variable?

**Comment by hydra** [comment]
@martinbudden I'd like to see this in cleanflight too please.  There is definate merit in the ability to keep an eye on stack usage.  even just so you can say to other developers 'make sure you code doesn't eat too much stack' and tell them how to check.

It might be worth breaking out the code in taskStackCheck into a stackCheck() method you can call from anywhere and then the taskStackCheck just calls that too.

**Comment by hydra** [comment]
and yes @digitalentity probably a some stack check state variable would be good so it's easier to reference when debugging.

**Comment by martinbudden** [comment]
@hydra , OK I'll do a PR for Cleanflight with the changes suggested.

**Comment by digitalentity** [comment]
Still some testing of the code is required.

**Comment by hydra** [comment]
keep us posted with regards to testing.

**Comment by digitalentity** [comment]
@martinbudden there is no code that will fill stack with 0xA5 at boot-up. Linker skript indicates fill for ELF but there is nothing in the non-eabi code that will pre-fill the stack.

**Comment by digitalentity** [comment]
Superseded by #642

**Pull request closed without merging**